File: web-studio/src/sign/proof.ts
/**
 * Cryptographic proof for a signature (optional "advanced" layer).
 *
 * Two independent facts are checked at verification time:
 *   1. Authenticity — does the Ed25519 signature verify against the embedded
 *      public key? (and, if a trusted key is supplied, does it match?)
 *   2. Integrity   — is the document still identical to what was signed?
 *
 * A signature is NEVER a qualified electronic signature. See la Documentation (§7).
 */

import { canonicalJSON, sha256Hex, nowIso } from "../format/canonical";
import type {
  EliumDocumentModel,
  EliumSignature,
  SignaturePlacement,
  SignatureProof,
  SignatureVerdict,
  SignatureVisual,
  SignerInfo,
} from "../format/types";
import { fingerprintOf, publicKeyHexFromPrivate, signMessage, verifyMessage } from "./keys";

/** Hash of the canonical document model (page settings + body). */
export async function computeContentHash(model: EliumDocumentModel): Promise<string> {
  return sha256Hex(canonicalJSON(model));
}

/**
 * The exact bytes that get signed — a stable, self-describing structure.
 * `signedPlacement`/`signedVisual` are included only when provided, so legacy
 * proofs (created before placement/visual were bound) reconstruct byte-identical.
 */
function toBeSigned(
  signatureId: string,
  signedContentHash: string,
  signer: SignerInfo,
  signedAt: string,
  placement?: SignaturePlacement,
  visual?: SignatureVisual,
): string {
  return canonicalJSON({
    v: 1,
    signatureId,
    signedContentHash,
    signer,
    signedAt,
    ...(placement ? { signedPlacement: placement } : {}),
    ...(visual ? { signedVisual: visual } : {}),
  });
}

export async function createProof(opts: {
  signatureId: string;
  model: EliumDocumentModel;
  signer: SignerInfo;
  privateKeyHex: string;
  /** Bind the signature's placement & appearance into the proof (recommended). */
  placement?: SignaturePlacement;
  visual?: SignatureVisual;
}): Promise<SignatureProof> {
  const publicKeyHex = await publicKeyHexFromPrivate(opts.privateKeyHex);
  const signedContentHash = await computeContentHash(opts.model);
  const signedAt = nowIso();
  const message = toBeSigned(opts.signatureId, signedContentHash, opts.signer, signedAt, opts.placement, opts.visual);
  const signatureHex = await signMessage(message, opts.privateKeyHex);

  return {
    alg: "ed25519",
    publicKeyHex,
    fingerprint: await fingerprintOf(publicKeyHex),
    contentHashAlg: "sha-256",
    signedContentHash,
    signatureHex,
    signedAt,
    ...(opts.placement ? { signedPlacement: opts.placement } : {}),
    ...(opts.visual ? { signedVisual: opts.visual } : {}),
    timestamp: {
      type: "local",
      at: signedAt,
      note: "Horodatage local non qualifié (pas d'autorité de temps).",
    },
  };
}

/**
 * @param signature the full signature (its `id` is part of the signed message)
 * @param model     the current document (to detect post-signature changes)
 * @param trustedKeyHex optional expected signer key for attribution
 */
export async function verifyProof(
  signature: EliumSignature,
  model: EliumDocumentModel,
  trustedKeyHex?: string,
): Promise<SignatureVerdict> {
  const proof = signature.proof;
  if (!proof) return "visual_only";

  // Reconstruct against the SNAPSHOTS embedded in the proof (present ⇒ v2, the
  // proof binds placement/visual; absent ⇒ legacy, it doesn't). Their presence
  // changes the signed bytes, so an attacker can neither strip nor forge them
  // without the signature failing.
  const message = toBeSigned(
    signature.id,
    proof.signedContentHash,
    signature.signer,
    proof.signedAt,
    proof.signedPlacement,
    proof.signedVisual,
  );
  const authentic = await verifyMessage(proof.signatureHex, message, proof.publicKeyHex);
  if (!authentic) return "invalid";

  if (trustedKeyHex && trustedKeyHex.trim().toLowerCase() !== proof.publicKeyHex.toLowerCase()) {
    return "unknown_key";
  }

  if ((await computeContentHash(model)) !== proof.signedContentHash) return "modified";
  // A v2 proof also detects the signature being moved or its appearance altered.
  if (proof.signedPlacement && canonicalJSON(signature.placement) !== canonicalJSON(proof.signedPlacement))
    return "modified";
  if (proof.signedVisual && canonicalJSON(signature.visual) !== canonicalJSON(proof.signedVisual)) return "modified";
  return "valid";
}

const VERDICT_LABELS: Record<SignatureVerdict, string> = {
  valid: "Signature valide",
  modified: "Document modifié après signature",
  invalid: "Signature invalide",
  unknown_key: "Clé inconnue (non vérifiée)",
  visual_only: "Signature visuelle uniquement",
};

export function verdictLabel(v: SignatureVerdict): string {
  return VERDICT_LABELS[v];
}

File: src/elium/format/proof.py
"""
Cryptographic signature proof (Python mirror of sign/proof.ts).

Interoperable with the web implementation: the signed message and content hash
use the same canonical JSON, so a proof created on one side verifies on the other.
"""

from __future__ import annotations

import hashlib
from typing import Any

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey

from elium.format.canonical import canonical_json, now_iso, sha256_hex

_TS_NOTE = "Horodatage local non qualifié (pas d'autorité de temps)."


def compute_content_hash(model: dict) -> str:
    return sha256_hex(canonical_json(model))


def _to_be_signed(
    signature_id: str,
    signed_hash: str,
    signer: dict,
    signed_at: str,
    placement: dict | None = None,
    visual: dict | None = None,
) -> str:
    """Canonical to-be-signed structure (mirror of proof.ts).

    `signedPlacement`/`signedVisual` are included only when provided, so legacy
    proofs (without them) reconstruct byte-identical.
    """
    payload: dict[str, Any] = {
        "v": 1,
        "signatureId": signature_id,
        "signedContentHash": signed_hash,
        "signer": signer,
        "signedAt": signed_at,
    }
    if placement is not None:
        payload["signedPlacement"] = placement
    if visual is not None:
        payload["signedVisual"] = visual
    return canonical_json(payload)


def generate_identity() -> dict[str, str]:
    priv = Ed25519PrivateKey.generate()
    priv_raw = priv.private_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PrivateFormat.Raw,
        encryption_algorithm=serialization.NoEncryption(),
    )
    pub_raw = priv.public_key().public_bytes(
        encoding=serialization.Encoding.Raw, format=serialization.PublicFormat.Raw
    )
    return {
        "privateKeyHex": priv_raw.hex(),
        "publicKeyHex": pub_raw.hex(),
        "fingerprint": hashlib.sha256(pub_raw).hexdigest(),
    }


def create_proof(
    signature_id: str,
    model: dict,
    signer: dict,
    private_key_hex: str,
    placement: dict | None = None,
    visual: dict | None = None,
) -> dict[str, Any]:
    priv = Ed25519PrivateKey.from_private_bytes(bytes.fromhex(private_key_hex))
    pub_raw = priv.public_key().public_bytes(
        encoding=serialization.Encoding.Raw, format=serialization.PublicFormat.Raw
    )
    public_key_hex = pub_raw.hex()
    signed_hash = compute_content_hash(model)
    signed_at = now_iso()
    message = _to_be_signed(signature_id, signed_hash, signer, signed_at, placement, visual)
    signature = priv.sign(message.encode("utf-8"))
    proof: dict[str, Any] = {
        "alg": "ed25519",
        "publicKeyHex": public_key_hex,
        "fingerprint": hashlib.sha256(pub_raw).hexdigest(),
        "contentHashAlg": "sha-256",
        "signedContentHash": signed_hash,
        "signatureHex": signature.hex(),
        "signedAt": signed_at,
    }
    if placement is not None:
        proof["signedPlacement"] = placement
    if visual is not None:
        proof["signedVisual"] = visual
    proof["timestamp"] = {"type": "local", "at": signed_at, "note": _TS_NOTE}
    return proof


def verify_proof(signature: dict, model: dict, trusted_key_hex: str | None = None) -> str:
    """Returns: valid | modified | invalid | unknown_key | visual_only."""
    proof = signature.get("proof")
    if not proof:
        return "visual_only"

    # Reconstruct against the snapshots embedded in the proof (present ⇒ v2, the
    # proof binds placement/visual; absent ⇒ legacy). Mirror of proof.ts.
    message = _to_be_signed(
        signature["id"],
        proof["signedContentHash"],
        signature.get("signer", {}),
        proof["signedAt"],
        proof.get("signedPlacement"),
        proof.get("signedVisual"),
    ).encode("utf-8")
    try:
        pub = Ed25519PublicKey.from_public_bytes(bytes.fromhex(proof["publicKeyHex"]))
        pub.verify(bytes.fromhex(proof["signatureHex"]), message)
    except (InvalidSignature, ValueError):
        return "invalid"

    if trusted_key_hex and trusted_key_hex.strip().lower() != proof["publicKeyHex"].lower():
        return "unknown_key"

    if compute_content_hash(model) != proof["signedContentHash"]:
        return "modified"
    # A v2 proof also detects the signature being moved or its appearance altered.
    if "signedPlacement" in proof and canonical_json(signature.get("placement")) != canonical_json(proof["signedPlacement"]):
        return "modified"
    if "signedVisual" in proof and canonical_json(signature.get("visual")) != canonical_json(proof["signedVisual"]):
        return "modified"
    return "valid"
